FEATURE: Add Fusion prototype for flash messages

Kleisli opened the pull request:
related: #3785 

**Review instructions**

The prototype can be used to render the flash messages in AFX like
```
<ul>
    <Neos.Fusion:FlashMessages severity="Error">
        <li>{message.message}</li>
    </Neos.Fusion:FlashMessages>
</ul>
```
but if there is no path /content to render the prototype returns the Messages as array and they can be processed further
```
flashMessages = Neos.Fusion:FlashMessages
renderer = Neos.Fusion:Loop {
    items={props.flashMessages}
    itemRenderer = My.Package:MyFlashMessageRenderer
} 
```
as Neos.Fusion:FlashMessages extends Neos.Fusion:Map, it inherits the iteration data feature
and as for Neos.Fusion:Loop the content between messages can be defined with the meta property `@glue`

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

Kleisli committed:
FEATURE 3785 add Fusion prototype for FlashMessages

Kleisli commented:
I just noticed the problem in the AFX example, when there are no messages we end up with an empty ul element...
I'll fix that and open a new PR.

mhsdesign commented:
hi thx for your pr: a few things noted:

- @propTypes doenst work for non components (and is also not used in the core - but that would be debateable)
- we need to adjust the docs and add the new Object (some rst file...)
- you dont need to close this pr - on contrary, just add more commits - also it tells a nice story
- i never thought of such conditonal logic with content so that looks fun - but:
- the conditional logic adds uneeded complexity (the inheritance etc) also by thinking about how often this flashmessage will be used and that presentation and integration should be seprated and that this is a non standart pattern - i just dont think that it is justified (sorry)

Kleisli commented:
@mhsdesign I agree with your points, let's see if I can reopen it and add my new version

Kleisli committed:
FEATURE 3785 add Fusion prototypes for FlashMessages

Kleisli commented:
I added the propTypes, because I was missing code comments for the current core prototypes, but now I saw that there are more comments in the lower fusion prototypes, so I used the same style for in-code documentation. but where should I add the documentation in an rst file, as you mentioned @mhsdesign ?

and I can't keep it simple.. reading your comment about separation of presentation and integration I thought there need to be an option to add flash messages in fusion code - at least for testing purposes 

this is what I also added as documentation in the fusion code
```
# returns the current flashMessages as array, passes then to the renderer or to the content
#
# Usage:
# flashMessages = Neos.Fusion:FlashMessages
# renderer = afx`
#   <Neos.Fusion:Loop items={props.flashMessages}>
#     {item.message}
#   </Neos.Fusion:Loop>
# `
#
# Usage with renderer:
# renderedFlashMessages = Neos.Fusion:FlashMessages {
#   renderer = My.Package:FlashMessageRenderer
# }
# renderer = afx`
#   {props.renderedFlashMessages}
# `
#
# AFX only:
# renderer = afx`
#   <Neos.Fusion:FlashMessages>
#     <Neos.Fusion:Loop items={flashMessages}>
#       {item.message}
#     </Neos.Fusion:Loop>
#   </Neos.Fusion:FlashMessages>
# `
```
```
# to pass additional flashMessages to Neos.Fusion:FlashMessages within the fusion code
#
# Usage:
# flashMessages = Neos.Fusion:FlashMessages {
#   additionalMessages {
#     testCase1 = Neos.Fusion:FlashMessages.Message {
#       message = 'Test'
#     }
#   }
# }
```

Kleisli committed:
fix documentation comment

mhsdesign commented:
https://github.com/neos/neos-development-collection/blob/27b14719c0b48d4d5c12041b6e5155e59bb74fdd/Neos.Neos/Documentation/References/NeosFusionReference.rst

Kleisli commented:
@mhsdesign fyi. I just started a topic on discuss to debate the proptypes :)
https://discuss.neos.io/t/documenting-fusion/6056

Kleisli commented:
> https://github.com/neos/neos-development-collection/blob/27b14719c0b48d4d5c12041b6e5155e59bb74fdd/Neos.Neos/Documentation/References/NeosFusionReference.rst

@mficzel is there a specific order of the Neos.Fusion objects or in other words, where should I add Neos.Fusion:Flashmessages? or would it even be ok to re-sort them alphabetically?

Kleisli committed:
satisfy StyleCI

Kleisli committed:
satisfy StyleCI

mhsdesign commented on Neos.Fusion/Resources/Private/Fusion/Root.fusion:
```diff
@@ -289,3 +290,76 @@ prototype(Neos.Fusion:GlobalCacheIdentifiers) < prototype(Neos.Fusion:DataStruct
+#     }
+#   }
+# }
+prototype(Neos.Fusion:FlashMessages.Message) < prototype(Neos.Fusion:DataStructure) {
```
i was wondering what the usecase for additionalMessages is ... if this is only for functional testing - we can easily mock the flashmessage service ;)

Kleisli replied:
the use case for this would not be functional testing, but testing the output of flashmessages in the presentation layer. e.g. with Sitegeist.Monocle, like
```
@styleguide {
  props {
    flashMessages {
      additionalMessages {
        testCase1 = Neos.Fusion:FlashMessages.Message {
        message = 'Test'
      }
    }
  }
}
```
inspired by your comment "that presentation and integration should be separated"

mhsdesign replied:
hmm - i would be in the opinion that a flashmessage dto should not reach the presentation layer (same like node) and be mapped to an array. That way we dont need to instantiate the objects for monocle and the component can render independent of the message dto

mhsdesign replied:
also it reduces complexity since we dont need the Flashmessages object and no additional flashmessages logic

mhsdesign commented on Neos.Fusion/Resources/Private/Fusion/Root.fusion:
```diff
@@ -289,3 +290,76 @@ prototype(Neos.Fusion:GlobalCacheIdentifiers) < prototype(Neos.Fusion:DataStruct
+#
+# AFX only:
+# renderer = afx`
+#   <Neos.Fusion:FlashMessages>
+#     <Neos.Fusion:Loop items={flashMessages}>
+#       {item.message}
+#     </Neos.Fusion:Loop>
+#   </Neos.Fusion:FlashMessages>
```
Nice i like that much better ;) Cool idea ^^

mhsdesign reviewed (commented):
Thx also for the discussion: https://discuss.neos.io/t/documenting-fusion/6056

Types in fusion cannot be discussed often enough ^^

mhsdesign commented:
With #3998 there might be little to no demand for this ... what do you say @mficzel ?

mficzel commented:
@mhsdesign This is imho still relevant 

FlashMessages can also be used in the Frontend. So this still makes sense to me. Also once render the main backend controller in Fusion we will need something like this there aswell.

mficzel reviewed (commented):
I assume you use the special cases for /content and /renderer to support afx which is a valid concern. However i suggest to use an eelHelper in this case which can be used in Fusion and afx.

```
<Neos.Fusion:Loop items={Neos.FlashMessages.get()} >
    ... custom rendering ... 
</ Neos.Fusion:Loop>
```

The eelHelper idea may be stupid and it may be hard to get access to the FlashMessageContainer from Eel. If this is the case i would somehow prefer a Fusion Prototype that only returns FlashMessages but am not strictly against including rendering.

mhsdesign commented:
I totally agree with @mficzel your currently proposed implementation would be a bit hard to maintain over time as its quite big.
A small eel helper would of course be the preferred way, as this would allow us also to separate integration and presentation easily.

(you might need some uglyish hack to get the controllerContext from within an eel helper, but thats okay ^^ (like try bootstrap->getActiveRequestHandler or something like this should lead the way ^^)

mhsdesign reviewed (changes requested):
I totally agree with @mficzel your currently proposed implementation would be a bit hard to maintain over time as its quite big.
A small eel helper would of course be the preferred way, as this would allow us also to separate integration and presentation easily.

(you might need some uglyish hack to get the controllerContext from within an eel helper, but thats okay ^^ (like try bootstrap->getActiveRequestHandler or something like this should lead the way ^^)

mficzel commented:
Looked further and the FlashmessageContainer is handled by the FlashMessageService that only needs the currentRequest. Since `request` is present in the FusionContext this should be very possible to implement ... 

```
<Neos.Fusion:Loop items={Neos.FlashMessages.forRequest(request)} >
    ... custom rendering ... 
</ Neos.Fusion:Loop>
```

See: 
- https://github.com/neos/flow-development-collection/blob/8.3/Neos.Flow/Classes/Mvc/Controller/ControllerContext.php#L123-L126
- https://github.com/neos/flow-development-collection/blob/8.3/Neos.Flow/Classes/Mvc/FlashMessage/FlashMessageService.php#L78-L82

mhsdesign commented:
Love it @mficzel  (seriously) no magic anymore ^^

Kleisli commented:
Sorry, I totally forgot about this PR of mine..

@mficzel I would like a solution like your first one, without having to pass the request.. from an architectural point of view I see that this is necessary, but as a developer it is annoying having to pass a fixed argument without having an actual choice what to pass..

But I think I have a solution for this: If we implement the Eel-Helper, we could add a prototype to Neos.Fusion like
```
prototype(Neos.Fusion:RenderFlashMessages) < prototype(Neos.Fusion:Component) {
  # severity of the messages to return (One of the Neos\Error\Messages\Message::SEVERITY_* constants)
  severity = null

  # whether the messages should be flushed after reading them, default is true
  flushMessages = true

  renderer = Neos.Fusion:Loop {
    items = ${FlashMessages.forRequest(request, props.severity, props.flushMessages)} 
    itemRenderer = ${item.message}
  }

}
```

and I suggest to add the FlashMessage Eel helper to Neos.Eel and omit the Neos namespace like for the other core Eel helpers in Neos-Eel, because FlashMessages can also be used in pure Flow applications.

mhsdesign commented:
> I would like a solution like your first one, without having to pass the request.. from an architectural point of view I see that this is necessary, but as a developer it is annoying having to pass a fixed argument without having an actual choice what to pass.

I would generally agree if this would be a really prominent use-case. BUT, as of now its basically impossible to render the flashmessages at all and thus nearly never done, so having a simple eel helper with no big abstraction would be already a great improvement.

To further abstract this away one can still write a custom fusion wrapping utility for your project.
